crates/provider-sdk/src/lib.rs

```diff
@@ -142,3 +150,107 @@ pub trait ProviderHandler: Sync {
+) -> ProviderInitResult<Vec<String>> {
+    let lattice_name = lattice_name.as_ref();
+    let wrpc_version = wrpc_version.as_ref();
+
+    let mut topics = Vec::new();
+    for (_, world) in resolve.worlds.iter() {
+        let world_name = &world.name;
+
+        let pkg_id = world.package.ok_or_else(|| {
+            ProviderInitError::Initialization(format!(
+                "unexpectedly missing package on with world [{world_name}]"
+            ))
+        })?;
+        let pkg = resolve.packages.get(pkg_id).ok_or_else(|| {
+            ProviderInitError::Initialization(format!(
+                "unexpectedly package with ID [{pkg_id:?}] in world [{world_name}]"
+            ))
+        })?;
+        let pkg_name = &pkg.name.name;
+        let ns = &pkg.name.namespace;
+
+        for (world_item, _) in world.exports.iter() {
+            match world_item {
+                wit_parser::WorldKey::Name(_) => continue,
+                wit_parser::WorldKey::Interface(iface_id) => {
+                    if let Some(iface) = resolve.interfaces.get(*iface_id) {
+                        let iface_name = iface.name.as_ref().ok_or_else(|| {
+                            ProviderInitError::Initialization(format!(
+                                "unexpectedly un-named iface ID [{iface_id:?}] in world [{world_name}]"
+                            ))
+                        })?;
+
+                        for (fn_name, _) in iface.functions.iter() {
+                            topics.push(format!(
+                                "{lattice_name}.wrpc.{wrpc_version}.{ns}:${pkg_name}/{iface_name}.{fn_name}"
+                            ));
+                        }
+                    }
+                }
+            }
+        }
+    }
+    Ok(topics)
+}
+
+/// An object that can serve as a way to look up the contents of files
+///
+/// Use of this trait should enable forward evolution of different file lookup
+/// methods without breaking provider_sdk
+#[async_trait]
+pub trait FileLookup {
+    /// Retrieve a single file
+    async fn get_file(&self, p: PathBuf) -> Option<&Vec<u8>>;
+}
+
+#[async_trait]
+impl FileLookup for HashMap<PathBuf, Vec<u8>> {
+    async fn get_file(&self, p: PathBuf) -> Option<&Vec<u8>> {
+        self.get(&p)
+    }
+}
+
+/// A trait for providers that are powered by WIT contracts
+///
+/// Providers are responsible for carrying the contents of their `wit`
+/// directories so they can be made available to code (ex. in `provider-sdk`)
+#[async_trait]
+pub trait WitContract {
+    /// Retrieve all WIT related files from the provider
+    ///
+    /// This is a lower level API to enable processing of rebuilding of the
+    /// `wit_parser::Resolve` structure, when processing WIT-related information
+    ///
+    /// Normally, the provider only has to implement this method, returning contents
+    /// of files in the WIT folder.
+    async fn get_wit_files(&self) -> ProviderInitResult<Box<dyn FileLookup + Send>> {
```

Comment:
I'd probably not provide a default implementation here since people really should be including the wit

Reply:
Oh ok that makes sense!